### ISSUE TITLE
fix -DNDEBUG warnings

### ISSUE THREAD

**Issue by cwsmith** [SCOREC/core#86]
We should fix the warnings that appear when we compile with `-DNDEBUG` and `-DSCOREC_CXX_WARNINGS=ON`.

```(void)foo;``` should do the trick; [as done here.](https://github.com/SCOREC/core/blob/5198f7de40b78bc96a79f688d7b2aaae61c614a6/lion/lionNoZLib.cc#L9-L15)

related to #81

**Comment by bgranzow** [comment]
Working through this it is tough for me to overstate how bad of an idea it seems to be to turn -DNDEBUG on

**Comment by bgranzow** [comment]
The proposed solution involves implementing two (or perhaps more to account for error handling) custom ASSERT macros, one that is always on and one that can be optionally turned off when NDEBUG is turned on. Using Albany as an example:

https://github.com/gahansen/Albany/blob/master/src/Albany_Utils.hpp#L202

**Comment by bgranzow** [comment]
This looks like it should be fixed as of this commit:

**Comment by ibaned** [comment]
In Albany, I was using `__VA_ARGS__` to make the verbose message optional. It looks like your macros expect the `...` to just be the condition being tested, so I would replace those and `__VA_ARGS__` with `cond` or something.

**Comment by bgranzow** [comment]
Done in 603033eb

The code now compiles with `-DNDEBUG=ON`.
Four macros have been defined:
```
PCU_ALWAYS_ASSERT(cond)
PCU_ALWAYS_ASSERT_VERBOSE(cond, ms)
PCU_DEBUG_ASSERT(cond)
PCU_DEBUG_ASSERT_VERBOSE(cond, msg)
```
All `assert`'s have been replaced with `PCU_ALWAYS_ASSERT()`.  As the name suggests, this assert is never disabled. Developers can now implement or revise asserts that should get disabled with `-DNDEBUG=ON` using the `PCU_DEBUG*` methods. The rationale for this is to minimize the impact of making "fast code" choices in the code base, so that the onus of making these decisions is placed on the developers who must carefully choose in which instances they should occur. This is in contrast to going through the entire code base and removing warnings=errors that are associated with unused variables when `-DNDEBUG=ON.`

tldr; asserts in the code do some pretty important stuff. you now have to choose where you want "fast code" to get executed with `-DNDEBUG=ON`

Additionally, the `*ASSERT_VERBOSE` options now provide developers the ability to print a user-friendly error message along with the assertion failure so that user's don't have to decipher cryptic messages like:
`r == n failed`
